[Pytorch Edge][tracing-based] reorganize model tracer dependency

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66237
* #64152
* #64087
* __->__ #63421

1. move 4 files to :
```
KernelDTypeTracer.h
KernelDTypeTracer.h
OperatorCallTracer.h
OperatorCallTracer.h
```
so it's visible in OSS.

2. Update the namespace to `torch::jit::mobile`
3. Add a `fb_xplat_cxx_library` `torch_model_tracer` with the source file list above.
4. update the `fb_xplat_cxx_library`  `model_tracer_lib` dependency on the new `torch_model_tracer` library



Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069)

Commit:
[Pytorch Edge][tracing-based] reorganize model tracer dependency



Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


1. move 4 files to :
```
KernelDTypeTracer.h
KernelDTypeTracer.h
OperatorCallTracer.h
OperatorCallTracer.h
```
so it's visible in OSS.

2. Update the namespace to `torch::jit::mobile`
3. Add a `fb_xplat_cxx_library` `torch_model_tracer` with the source file list above.
4. update the `fb_xplat_cxx_library`  `model_tracer_lib` dependency on the new `torch_model_tracer` library



Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


1. move 4 files to :
```
KernelDTypeTracer.h
KernelDTypeTracer.h
OperatorCallTracer.h
OperatorCallTracer.h
```
so it's visible in OSS.

2. Update the namespace to `torch::jit::mobile`
3. Add a `fb_xplat_cxx_library` `torch_model_tracer` with the source file list above.
4. update the `fb_xplat_cxx_library`  `model_tracer_lib` dependency on the new `torch_model_tracer` library



Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


1. move 4 files to :
```
KernelDTypeTracer.h
KernelDTypeTracer.h
OperatorCallTracer.h
OperatorCallTracer.h
```
so it's visible in OSS.

2. Update the namespace to `torch::jit::mobile`
3. Add a `fb_xplat_cxx_library` `torch_model_tracer` with the source file list above.
4. update the `fb_xplat_cxx_library`  `model_tracer_lib` dependency on the new `torch_model_tracer` library



Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


1. move 4 files to :
```
KernelDTypeTracer.h
KernelDTypeTracer.h
OperatorCallTracer.h
OperatorCallTracer.h
```
so it's visible in OSS.

2. Update the namespace to `torch::jit::mobile`
3. Add a `fb_xplat_cxx_library` `torch_model_tracer` with the source file list above.
4. update the `fb_xplat_cxx_library`  `model_tracer_lib` dependency on the new `torch_model_tracer` library



Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


1. move 4 files to :
```
KernelDTypeTracer.h
KernelDTypeTracer.h
OperatorCallTracer.h
OperatorCallTracer.h
```
so it's visible in OSS.

2. Update the namespace to `torch::jit::mobile`
3. Add a `fb_xplat_cxx_library` `torch_model_tracer` with the source file list above.
4. update the `fb_xplat_cxx_library`  `model_tracer_lib` dependency on the new `torch_model_tracer` library



Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


1. move 4 files to :
```
KernelDTypeTracer.h
KernelDTypeTracer.h
OperatorCallTracer.h
OperatorCallTracer.h
```
so it's visible in OSS.

2. Update the namespace to `torch::jit::mobile`
3. Add a `fb_xplat_cxx_library` `torch_model_tracer` with the source file list above.
4. update the `fb_xplat_cxx_library`  `model_tracer_lib` dependency on the new `torch_model_tracer` library



Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


1. move 4 files to :
```
KernelDTypeTracer.h
KernelDTypeTracer.h
OperatorCallTracer.h
OperatorCallTracer.h
```
so it's visible in OSS.

2. Update the namespace to `torch::jit::mobile`
3. Add a `fb_xplat_cxx_library` `torch_model_tracer` with the source file list above.
4. update the `fb_xplat_cxx_library`  `model_tracer_lib` dependency on the new `torch_model_tracer` library



Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069)

[ghstack-poisoned]

malfet commented on torch/csrc/jit/mobile/model_tracer/OperatorCallTracer.h:
```diff
@@ -0,0 +1,48 @@
+ *
+ */
+struct OperatorCallTracer final {
+  static std::set<std::string> called_operators_;
```
Does this set really needs to be ordered? I.e. wouldn't `std::unordered_set` be faster?

JacobSzwejbka replied:
Its expected for these traced operators to be serialized so its convenient for the ordering to always be consistent. The efficiency difference should be small, the majority of models in my experience have ~100 operators.

malfet commented on torch/csrc/jit/mobile/model_tracer/OperatorCallTracer.h:
```diff
@@ -0,0 +1,48 @@
+  static std::set<std::string> called_operators_;
+  at::CallbackHandle handle_;
+
+  OperatorCallTracer() {
```
Is `OperatorCallTracer` supposed to be a singleton? If so, wouldn't it be better to make its constructor private?

malfet commented on torch/csrc/jit/mobile/model_tracer/KernelDTypeTracer.h:
```diff
@@ -0,0 +1,56 @@
+   * std::set<std::string> is the collection of dtypes for which we need to
+   * generate code for the said kernel tag.
+   */
+  typedef std::map<std::string, std::set<std::string>> kernel_tags_type;
```
Is order of keys in a map important? If not, please use unordered map and unordered set
```suggestion
  typedef std::unordered_map<std::string, std::unordered_set<std::string>> kernel_tags_type;
```

cccclai commented:
Thanks Nikita for the feedback. Let me have a separate pr on top of the pr stack to address them.

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


1. move 4 files to :
```
KernelDTypeTracer.h
KernelDTypeTracer.h
OperatorCallTracer.h
OperatorCallTracer.h
```
so it's visible in OSS.

2. Update the namespace to `torch::jit::mobile`
3. Add a `fb_xplat_cxx_library` `torch_model_tracer` with the source file list above.
4. update the `fb_xplat_cxx_library`  `model_tracer_lib` dependency on the new `torch_model_tracer` library



Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


1. move 4 files to :
```
KernelDTypeTracer.h
KernelDTypeTracer.h
OperatorCallTracer.h
OperatorCallTracer.h
```
so it's visible in OSS.

2. Update the namespace to `torch::jit::mobile`
3. Add a `fb_xplat_cxx_library` `torch_model_tracer` with the source file list above.
4. update the `fb_xplat_cxx_library`  `model_tracer_lib` dependency on the new `torch_model_tracer` library



Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] reorganize model tracer dependency"


1. move 4 files to :
```
KernelDTypeTracer.h
KernelDTypeTracer.h
OperatorCallTracer.h
OperatorCallTracer.h
```
so it's visible in OSS.

2. Update the namespace to `torch::jit::mobile`
3. Add a `fb_xplat_cxx_library` `torch_model_tracer` with the source file list above.
4. update the `fb_xplat_cxx_library`  `model_tracer_lib` dependency on the new `torch_model_tracer` library



Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069)

[ghstack-poisoned]
